hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/ReadJdbcP.java

```diff
@@ -0,0 +1,145 @@
+            @Nonnull ResultSetForPartitionFunction resultSetFn,
+            @Nonnull DistributedFunction<ResultSet, T> mapOutputFn
+    ) {
+        this.connectionSupplier = connectionSupplier;
+        this.resultSetFn = resultSetFn;
+        this.mapOutputFn = mapOutputFn;
+        setCooperative(false);
+    }
+
+    /**
+     * Private API, use {@link SourceProcessors#readJdbcP}.
+     */
+    public static <T> ProcessorSupplier supplier(
+            @Nonnull DistributedSupplier<Connection> connectionSupplier,
+            @Nonnull ResultSetForPartitionFunction resultSetFn,
+            @Nonnull DistributedFunction<ResultSet, T> mapOutputFn
+    ) {
+        return ProcessorSupplier.of(() -> new ReadJdbcP<>(connectionSupplier, resultSetFn, mapOutputFn));
+    }
+
+    public static <T> ProcessorSupplier supplier(
+            @Nonnull String connectionURL,
+            @Nonnull String query,
+            @Nonnull DistributedFunction<ResultSet, T> mapOutputFn
+    ) {
+        return ProcessorSupplier.of(() -> new ReadJdbcP<>(
+                () -> uncheckCall(() -> DriverManager.getConnection(connectionURL)),
+                resultSetFunction(query), mapOutputFn));
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        connection = connectionSupplier.get();
+        this.parallelism = context.totalParallelism();
+        this.index = context.globalProcessorIndex();
+    }
+
+    @Override
+    public boolean complete() {
+        if (traverser == null) {
+            resultSet = resultSetFn.resultSet(connection, parallelism, index);
+            traverser = ((Traverser<ResultSet>) () -> uncheckCall(() -> resultSet.next() ? resultSet : null))
+                    .map(mapOutputFn);
+        }
+        return emitFromTraverser(traverser);
+    }
+
+    @Override
+    public void close() throws Exception {
+        Exception resultSetException = null;
+        Exception statementException = null;
+        if (resultSet != null) {
+            Statement statement = resultSet.getStatement();
+            resultSetException = close(resultSet);
+            if (statement != null) {
+                statementException = close(statement);
+            }
+        }
+        if (connection != null) {
+            connection.close();
+        }
+        if (resultSetException != null) {
+            throw resultSetException;
+        }
+        if (statementException != null) {
+            throw statementException;
+        }
+    }
+
+    private static Exception close(AutoCloseable closeable) {
+        try {
+            closeable.close();
+        } catch (Exception e) {
+            return e;
+        }
+        return null;
+    }
+
+    private static ResultSetForPartitionFunction resultSetFunction(String query) {
+        return (connection, parallelism, index) -> {
+            PreparedStatement statement = uncheckCall(() -> connection.prepareStatement(query));
```

Comment:
we can assert `parallelism == 1`, otherwise it would emit duplicates.

Reply:
we use `ProcessorMetaSupplier.forceTotalParallelismOne` on `SourceProcessors.readJdbcP`, isn't it enough ?

Reply:
Yes, it is, but it's in different place. It's safety if anything goes wrong...
Btw, why do we return PS from the supplier and then wrap it in another place? It requires to be wrapped, why not return PMS right away?